### PR TITLE
feat(TikTok): added thumbnails & fixed profile title

### LINKED ISSUE #5200: Tiktok | tiktok.com
### Description

Not showing what im watching on tiktok

### Steps to Reproduce

1. Watch a random video, from a random tiktok profile
2. Go back to discord
3. See the status activity on discord

### Screenshots

image

### Expected Behavior

i was expected to show what im watching on tiktok

### Actual Behavior

Viewing profile of: undefined (@undefined)

### Page URL

https://www.tiktok.com/@studiodmofficial/video/7040533234383375622?is_copy_url=1&is_from_webapp=v1

### Operating System (OS)

Windows 10

### Browser

Brave

### Browser Version

Version 1.33.106 Chromium: 96.0.4664.110 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by DarkVillager**
Resolves #5200
**Describe the changes in this pull request:**
- Fixed user's username, etc not showing
- Fixed video paused & timestamp not working
- Added large imagekey when visiting someone's profile

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

</details>

### CODE REVIEW

**Comment by DarkVillager** [comment]
Will have to fix the rest of the presence oopsie will be back

File: websites/T/TikTok/presence.ts [hunk 1]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
+						.replace("window['SIGI_STATE']=", "")
+				).UserModule.users[id2].nickname
+			}`;
+		}
 		if (pageType === "video") {
 			//Video
-
-			const video = document.querySelector<HTMLVideoElement>(".video-player");
+			const time = document
+				.querySelectorAll(`[class*="-DivSeekBarTimeContainer"]`)[0]
```
Comment (DarkVillager) [inline websites/T/TikTok/presence.ts:66]: This doesn't work obv:
`.querySelectorAll("[class*="-DivSeekBarTimeContainer"]")[0]`

File: websites/T/TikTok/presence.ts [hunk 2]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
+				.innerHTML.split(`/`);
```
Comment (DarkVillager) [inline websites/T/TikTok/presence.ts:67]: It want me to use the following even tho it doesn't work:
`.innerHTML.split("\");` 

This is what I used before but that isn't alowed either:
`.innerHTML.split("/\");`

File: websites/T/TikTok/presence.ts (hunk 1 above)
Comment (ndbiaw) [inline websites/T/TikTok/presence.ts:66]:
```suggestion
			const time = document
				.querySelector(`[class*="-DivSeekBarTimeContainer"]`)
```

File: websites/T/TikTok/presence.ts (hunk 2 above)
Comment (ndbiaw) [inline websites/T/TikTok/presence.ts:67]:
```suggestion
				.innerHTML.split("/");
```

**Review by nayumiDEV — COMMENTED** [review]
it is plain text...

File: websites/T/TikTok/presence.ts (hunk 1 above)
Comment (ndbiaw) [reply websites/T/TikTok/presence.ts:66]: why?

File: websites/T/TikTok/presence.ts (hunk 2 above)
Comment (DarkVillager) [reply websites/T/TikTok/presence.ts:67]: its a \ not a /

File: websites/T/TikTok/presence.ts [hunk 3]
```diff
@@ -90,12 +114,10 @@ presence.on("UpdateData", async () => {
 				}
 			];
 		} else {
-			presenceData.details = (await strings).viewProfile;
-			presenceData.state = `${document
-				.querySelector(".share-sub-title")
-				?.textContent.trim()} (@${document
-				.querySelector(".share-title")
-				?.textContent.trim()})`;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:98]:
```ts
document.querySelector("[data-e2e=user-subtitle]").textContent.trim();
document.querySelector("[data-e2e=user-title]").textContent.trim();
```

File: websites/T/TikTok/presence.ts [hunk 4]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
+				.textContent.trim();
+			presenceData.state = `@(${id2}) ${
+				JSON.parse(
+					document
+						.querySelector("#sigi-persisted-data")
+						.innerHTML.slice(0, -486)
+						.replace("window['SIGI_STATE']=", "")
+				).UserModule.users[id2].nickname
+			}`;
+		}
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:62]:
```suggestion
```

File: websites/T/TikTok/presence.ts [hunk 5]
```diff
@@ -74,9 +101,6 @@ presence.on("UpdateData", async () => {
 			//Live
 			delete presenceData.startTimestamp;
 			presenceData.details = document.querySelector(".live-title")?.textContent;
-			presenceData.state = `@${
-				document.querySelector(".user-uniqueId")?.textContent
-			}`;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:79]:
```ts
document.querySelector("[data-e2e=user-profile-uid]").textContent;
```

File: websites/T/TikTok/presence.ts [hunk 6]
```diff
@@ -74,9 +101,6 @@ presence.on("UpdateData", async () => {
 			//Live
 			delete presenceData.startTimestamp;
 			presenceData.details = document.querySelector(".live-title")?.textContent;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:103]:
```ts
document.querySelector("[data-e2e=user-profile-live-title]");
```

File: websites/T/TikTok/presence.ts [hunk 7]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
+				).UserModule.users[id2].nickname
+			}`;
+		}
 		if (pageType === "video") {
 			//Video
-
-			const video = document.querySelector<HTMLVideoElement>(".video-player");
+			const time = document
+				.querySelector('[class*="-DivSeekBarTimeContainer"]')
+				.innerHTML.split("/");
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:67]:
```ts
document.querySelector("[class*=DivVideoWrapper] video")
```

File: websites/T/TikTok/presence.ts [hunk 8]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
 
 			delete presenceData.startTimestamp;
 			presenceData.details =
 				document.querySelector(".video-meta-title:nth-child(1)")
 					?.firstElementChild?.textContent ??
 				document.querySelector(".tt-video-meta-caption")?.firstElementChild
 					?.textContent;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:74]: These don't exist anymore, so either change it to use the `tiktok.viewing` string or use the element below.
```ts
document.querySelector("[data-e2e=browse-video-desc]");
```

File: websites/T/TikTok/presence.ts [hunk 9]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
-			presenceData.state = `@${
-				document.querySelector(".user-username")?.textContent ??
-				document.querySelector(".author-uniqueId")?.textContent
-			}`;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:59]:
```ts
document.querySelector("[data-e2e=browse-username]").textContent; // UID
document.querySelector("[data-e2e=browser-nickname]").childNodes[0].textContent; // Username
```

File: websites/T/TikTok/presence.ts [hunk 10]
```diff
@@ -41,25 +41,52 @@ presence.on("UpdateData", async () => {
...
+			if (
+				!document.querySelector(
+					"#app > div.tiktok-19fglm-DivBodyContainer.etsvyce0 > div.tiktok-7t2h2f-DivBrowserModeContainer.e1xqvjno0 > div.tiktok-5uccoo-DivVideoContainer.e1xqvjno27 > svg.tiktok-i8t918-SvgPlayIcon.e1xqvjno10"
+				)
+			) {
+				[, presenceData.endTimestamp] = presence.getTimestamps(
+					presence.timestampFromFormat(time[0]),
+					presence.timestampFromFormat(time[1])
+				);
+			}
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:89]:
```suggestion
			presenceData.smallImageKey = video.paused ? "pause" : "play";
			if (!video.paused)
				presenceData.endTimestamp = presence.getTimestampsfromMedia(video)[1];
```

File: websites/T/TikTok/presence.ts (hunk 9 above)
Comment (DarkVillager) [reply websites/T/TikTok/presence.ts:59]: This doesn't work?

File: websites/T/TikTok/dist/metadata.json [hunk 11]
```diff
@@ -21,16 +21,12 @@
...
 	"logo": "https://i.imgur.com/KBAjCOX.png",
 	"thumbnail": "https://i.imgur.com/QAxUXB4.png",
 	"color": "#000000",
 	"category": "socials",
-	"tags": [
-		"tiktok",
-		"video",
-		"social"
-	],
+	"tags": ["tiktok", "video", "social"],
```
Comment (EncryptedDev) [inline websites/T/TikTok/dist/metadata.json:29]: Revert this change

File: websites/T/TikTok/presence.ts [hunk 12]
```diff
@@ -90,12 +103,14 @@ presence.on("UpdateData", async () => {
...
+			presenceData.state = `@${
+				document.querySelector("[data-e2e=user-title]").textContent
+			} (${document.querySelector("[data-e2e=user-subtitle]").textContent})`;
+
+			presenceData.largeImageKey = document.querySelector<HTMLMetaElement>(
+				'meta[property="og:image"]'
+			).content;
 			presenceData.buttons = [
 				{
 					label: (await strings).buttonViewProfile,
```
Comment (EncryptedDev) [inline websites/T/TikTok/presence.ts:116]: You should migrate to use the strings type like in Netflix presence etc

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Hi, just a few more requests here; let me know if you need anything

File: websites/T/TikTok/presence.ts (hunk 12 above)
Comment (DarkVillager) [reply websites/T/TikTok/presence.ts:116]: Will check netflix presence’s code out

Comment (EncryptedDev) [reply websites/T/TikTok/presence.ts:116]: It is already look how the strings are typed

Comment (DarkVillager) [reply websites/T/TikTok/presence.ts:116]: so huh wdym then

Comment (EncryptedDev) [reply websites/T/TikTok/presence.ts:116]: Use the type here, this removes the need of `(await strings).` constantly
https://github.com/PreMiD/Presences/pull/5833/files/84f76ed9d1d61d4577739648bdfae228b99f0ac9#diff-564cbd292c5b92c85f3072d482a5c5cb476da7ac3e930559ec99470f96c2b7e1R16

File: websites/T/TikTok/dist/metadata.json (hunk 11 above)
Comment (DarkVillager) [reply websites/T/TikTok/dist/metadata.json:29]: Prettier auto formats it like this?

Comment (EncryptedDev) [reply websites/T/TikTok/dist/metadata.json:29]: Run SE instead. SE should have the preferred format on the repo; you wouldn’t have needed to make the change if it was preferred

File: websites/T/TikTok/presence.ts [hunk 13]
```diff
@@ -28,38 +28,42 @@ presence.on("UpdateData", async () => {
...
+				document.querySelector("[data-e2e=video-author-nickname]")
+					?.childNodes[0]?.textContent ??
+				document.querySelector("[data-e2e=browse-nickname]")?.textContent
+			})`;
 			presenceData.smallImageKey = video.paused ? "pause" : "play";
+			presenceData.largeImageKey = document
+				.querySelector(
+					"#app > div.tiktok-19fglm-DivBodyContainer.etsvyce0 > div.tiktok-yp78ys-DivMainContainer.erck9ax0 > div:nth-child(1) > div:nth-child(1) > a > div > span > img"
+				)
+				?.getAttribute("src");
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:64]:
```ts
document.querySelector<HTMLImageElement>("[data-e2e=browse-user-avatar] img[class*=-ImgAvatar]")?.src ?? "tiktok";
```

**Review by Slowlife01 — COMMENTED** [review]
Update your screenshots too

File: websites/T/TikTok/presence.ts (hunk 13 above)
Comment (Slowlife01) [reply websites/T/TikTok/presence.ts:64]: Also, it'd be good if you add setting for this.

**Comment by DarkVillager** [comment]
> Update your screenshots too

Done.

File: websites/T/TikTok/presence.ts [hunk 14]
```diff
@@ -18,101 +19,142 @@ async function getStrings() {
...
-	if (oldLang !== newLang || !strings) {
-		oldLang = newLang;
-		strings = await getStrings();
-	}
+		[, page, pageType] = location.pathname.split("/"),
+		[buttons, covers] = await Promise.all([
+			presence.getSetting<boolean>("buttons"),
+			presence.getSetting<boolean>("covers")
+		]);
+	strings ??= await newStrings;
```
Comment (Slowlife01) [inline websites/T/TikTok/presence.ts:36]: Are you sure this will get updated

**Review by Slowlife01 — CHANGES_REQUESTED** [review]

File: websites/T/TikTok/presence.ts (hunk 14 above)
Comment (DarkVillager) [reply websites/T/TikTok/presence.ts:36]: yes

Comment (Slowlife01) [reply websites/T/TikTok/presence.ts:36]: Well, I'm not sure if you misunderstood me. But it won't get updated if you change a language. So just revert this change.
